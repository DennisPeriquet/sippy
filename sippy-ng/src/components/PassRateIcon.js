import { Tooltip } from '@material-ui/core'
import ArrowDownwardRoundedIcon from '@material-ui/icons/ArrowDownwardRounded'
import ArrowUpwardRoundedIcon from '@material-ui/icons/ArrowUpwardRounded'
import React from 'react'
import SyncAltRoundedIcon from '@material-ui/icons/SyncAltRounded'

/**
 * PassRateIcon returns an up, down, or side-to-side arrows
 * indicating whether something improved, regressed, or stayed
 * the same.
 */
export default function PassRateIcon(props) {
  let icon = ''

  if (Math.abs(props.improvement) <= 2) {
    icon = (
      <SyncAltRoundedIcon
        data-icon="SyncAltRoundedIcon"
        style={{ color: 'grey' }}
      />
    )
  } else if (props.improvement >= 2) {
    icon = (
      <ArrowUpwardRoundedIcon
        data-icon="ArrowUpwardRoundedIcon"
        style={{
          stroke: props.inverted ? 'darkred' : 'green',
          strokeWidth: 3,
          color: props.inverted ? 'darkred' : 'green',
        }}
      />
    )
  } else {
    icon = (
      <ArrowDownwardRoundedIcon
        data-icon="ArrowDownwardRoundedIcon"
        style={{
          stroke: props.inverted ? 'green' : 'darkred',
          strokeWidth: 3,
          color: props.inverted ? 'green' : 'darkred',
        }}
      />
    )
  }

  if (props.tooltip) {
    return <Tooltip title={props.improvement.toFixed(2) + '%'}>{icon}</Tooltip>
  } else {
    return icon
  }
}

PassRateIcon.defaultProps = {
  inverted: false,
  tooltip: false,
}
