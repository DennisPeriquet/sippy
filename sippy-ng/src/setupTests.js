/** @jest-environment setup-polly-jest/jest-environment-node */

import '@testing-library/jest-dom'
import Enzyme from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import fetch from 'node-fetch'
import path from 'path'
import { setupPolly } from 'setup-polly-jest'
import toDiffableHtml from 'diffable-html'

// https://github.com/mui-org/material-ui/issues/21293
export const withoutMuiID = (wrapper) => toDiffableHtml(wrapper.html()
  .replace(/id="mui-[0-9]*"/g, '')
  .replace(/aria-labelledby="(mui-[0-9]* *)*"/g, ''))

Enzyme.configure({ adapter: new Adapter() })

// To stub out fetch(), we need to use node-fetch and this line:
global.fetch = fetch

// See: https://github.com/vuejs/vue-test-utils/issues/974
global.requestAnimationFrame = cb => cb()

// When injecting multiple filters, material table complains.
const originalError = console.error.bind(console.error)

global.console.error = (log) => {
  if (!log.toString().includes('Warning: Failed')) {
    originalError(log.toString())
  }
}

// Set API URL for Sippy
process.env.REACT_APP_API_URL = 'http://localhost:8080'

// Default PollyJS mode is replay, set to record to record
// new API calls.
if (!process.env.POLLY_MODE) {
  process.env.POLLY_MODE = 'replay'
}

export const setupDefaultPolly = () => {
  const context = setupPolly({
    mode: process.env.POLLY_MODE,
    adapters: [require('@pollyjs/adapter-node-http')],
    persister: require('@pollyjs/persister-fs'),
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, '../__recordings__')
      }
    }
  })
  return context
}

export const expectLoadingPage = (wrapper) => expect(wrapper.find('p').contains('Loading...'))

jest.mock('react-chartjs-2', () => ({
  Bar: () => null,
  Chart: () => null,
  ChartComponent: () => null,
  Doughnut: () => null,
  Line: () => null,
  PolarArea: () => null
}))
