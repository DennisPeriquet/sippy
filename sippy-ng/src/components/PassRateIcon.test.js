import { act } from '@testing-library/react'
import { mount } from 'enzyme'
import { withoutMuiID } from '../setupTests'
import PassRateIcon from './PassRateIcon'
import React from 'react'

jest.useRealTimers()

describe(PassRateIcon, () => {
  it('shows sync icon for nearly no change', async () => {
    let wrapper
    await act(async () => {
      wrapper = mount(<PassRateIcon improvement={1} />)
    })

    expect(wrapper.find('svg[data-icon="SyncAltRoundedIcon"]').exists()).toBe(
      true
    )
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })

  it('shows up arrow for improvement', async () => {
    let wrapper
    await act(async () => {
      wrapper = mount(<PassRateIcon improvement={3} />)
    })

    expect(
      wrapper.find('svg[data-icon="ArrowUpwardRoundedIcon"]').exists()
    ).toBe(true)
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })

  it('shows down arrow for regression', async () => {
    let wrapper
    await act(async () => {
      wrapper = mount(<PassRateIcon improvement={-3} />)
    })

    expect(
      wrapper.find('svg[data-icon="ArrowDownwardRoundedIcon"]').exists()
    ).toBe(true)
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })

  it('renders tooltip', async () => {
    let wrapper
    await act(async () => {
      wrapper = mount(<PassRateIcon improvement={-3.9999999} tooltip={true} />)
    })

    expect(wrapper.exists()).toBe(true)
    expect(wrapper.find('ForwardRef(Tooltip)').props().title).toBe('-4.00%')
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })
})
