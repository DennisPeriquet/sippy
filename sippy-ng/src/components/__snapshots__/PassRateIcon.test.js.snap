// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PassRateIcon renders tooltip 1`] = `
<svg class="MuiSvgIcon-root"
     focusable="false"
     viewbox="0 0 24 24"
     aria-hidden="true"
     data-icon="ArrowDownwardRoundedIcon"
     style="stroke: darkred; stroke-width: 3; color: darkred;"
     title="-4.00%"
>
  <path d="M11 5v11.17l-4.88-4.88c-.39-.39-1.03-.39-1.42 0-.39.39-.39 1.02 0 1.41l6.59 6.59c.39.39 1.02.39 1.41 0l6.59-6.59c.39-.39.39-1.02 0-1.41a.9959.9959 0 00-1.41 0L13 16.17V5c0-.55-.45-1-1-1s-1 .45-1 1z">
  </path>
</svg>
`;

exports[`PassRateIcon shows down arrow for regression 1`] = `
<svg class="MuiSvgIcon-root"
     focusable="false"
     viewbox="0 0 24 24"
     aria-hidden="true"
     data-icon="ArrowDownwardRoundedIcon"
     style="stroke: darkred; stroke-width: 3; color: darkred;"
>
  <path d="M11 5v11.17l-4.88-4.88c-.39-.39-1.03-.39-1.42 0-.39.39-.39 1.02 0 1.41l6.59 6.59c.39.39 1.02.39 1.41 0l6.59-6.59c.39-.39.39-1.02 0-1.41a.9959.9959 0 00-1.41 0L13 16.17V5c0-.55-.45-1-1-1s-1 .45-1 1z">
  </path>
</svg>
`;

exports[`PassRateIcon shows sync icon for nearly no change 1`] = `
<svg class="MuiSvgIcon-root"
     focusable="false"
     viewbox="0 0 24 24"
     aria-hidden="true"
     data-icon="SyncAltRoundedIcon"
     style="color: grey;"
>
  <path d="M21.65 7.65l-2.79-2.79c-.32-.32-.86-.1-.86.35V7H4c-.55 0-1 .45-1 1s.45 1 1 1h14v1.79c0 .45.54.67.85.35l2.79-2.79c.2-.19.2-.51.01-.7zM2.35 16.35l2.79 2.79c.32.32.86.1.86-.35V17h14c.55 0 1-.45 1-1s-.45-1-1-1H6v-1.79c0-.45-.54-.67-.85-.35l-2.79 2.79c-.2.19-.2.51-.01.7z">
  </path>
</svg>
`;

exports[`PassRateIcon shows up arrow for improvement 1`] = `
<svg class="MuiSvgIcon-root"
     focusable="false"
     viewbox="0 0 24 24"
     aria-hidden="true"
     data-icon="ArrowUpwardRoundedIcon"
     style="stroke: green; stroke-width: 3; color: green;"
>
  <path d="M13 19V7.83l4.88 4.88c.39.39 1.03.39 1.42 0 .39-.39.39-1.02 0-1.41l-6.59-6.59a.9959.9959 0 00-1.41 0l-6.6 6.58c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L11 7.83V19c0 .55.45 1 1 1s1-.45 1-1z">
  </path>
</svg>
`;
